swarm: emit PeerConnectedness event from swarm instead of from hosts

marten-seemann opened the pull request:
Fixes #1571.

This was surprisingly easy. Unfortunately, I can't find a way to test the deduplication (that no event is emitted when a second connection is dialed), since the swarm is too good at preventing dials to hosts we're already connected to.

vyzo commented on p2p/host/basic/basic_host.go:
```diff
@@ -112,6 +112,9 @@ var _ host.Host = (*BasicHost)(nil)
 // HostOpts holds options that can be passed to NewHost in order to
 // customize construction of the *BasicHost.
 type HostOpts struct {
+	// EventBus sets the event bus. Will construct a new event bus if omitted.
```
I feel uneasy about this.
Maybr we should add the EventBus method to the Network interface and forward from the host so that wr cant end up with two of them.

marten-seemann replied:
Not really how this could happen, but we can make this change.

Can we defer this until we consolidated all the other repos as well (#1566)? It's happening in the next release.

Stebalien replied:
Another option would be to just make this entire thing private.

marten-seemann replied:
I'm not a big fan of having both a `Config` and `HostOpts`. I expect us to merge these two (or refactor them) in some other way when we start using Fx for service construction (#1993). I suggest not making any big changes in this PR.

MarcoPolo commented on p2p/net/swarm/swarm.go:
```diff
@@ -310,6 +321,13 @@ func (s *Swarm) addConn(tc transport.CapableConn, dir network.Direction) (*Conn,
 	})
 	c.notifyLk.Unlock()
 
+	if isFirstConn {
```
Are we sure we want this outside of the `notifyLk`? That changes the behavior from what it used to be (inside the notifyAll).

Stebalien replied:
Yeah, I think this needs to be inside. We need to make sure that the disconnect comes after the connect event.

Stebalien replied:
Hm. Actually, I'm not sure if this lock helps us here. We need to make this guarantee across all connections.

MarcoPolo commented on p2p/net/swarm/swarm.go:
```diff
@@ -588,9 +607,17 @@ func (s *Swarm) removeConn(c *Conn) {
+	s.conns.Unlock()
+
+	if disconnected {
+		s.emitter.Emit(event.EvtPeerConnectednessChanged{
```
What's the reason to do this here instead of near the old place where this happened: https://github.com/libp2p/go-libp2p/blob/marco/1575/p2p/net/swarm/swarm_conn.go#L85 ?

Stebalien replied:
Probably to ensure that we only emit the event on the last disconnect. But it does look like there's a bit of a race and we can emit a disconnect after a connect (looks like we already _have_ that race, actually).

Stebalien commented:
So, I think the only way to sanely do this is to write something like: https://github.com/ipfs/go-bitswap/pull/565/files. We probably don't want to block some global queue while sending events.

That is:

1. On connect/disconnect, mark the peer as connected/disconnected in some tracker. Add the peer to some "todo" queue, and tell the background process that it needs to wake up.
2. In a background event loop, fire connect/disconnect events.

Even if this background process falls behind (someone is blocking events), that's fine. We'll just always fire the "last" event. This _does_ mean we can get unmatched connected/disconnected events, but, IMO, that's _fine_.

I.e.:

1. Connect.
2. Disconnect.
3. Connect.
4. Disconnect.

May be seen as:

1. Connect.
2. Connect.
3. Disconnect.

Or

1. Disconnect.
2. Disconnect.

Or

1. Connect.
2. Disconnect.

Or

1. Disconnect.

But never:

1. Connect.
2. Disconnect.
3. Connect.

Importantly, the _final_ state is always accurate, which is what users care about. At the moment, it's possible for the _last_ event fired to be "connect", which could cause memory leaks.

marten-seemann commented:
Rebased this PR on top of the current master.

I think emitting the event while holding the `conns` mutex should be fine:
* It's correct since the the `swarm.conns.m[peer.ID]` is the source of truth: Once we're down to 0 connections in this list, we're officially disconnected from the peer.
* It's ok performance-wise, because emitting events is cheap, as long as the subscribers consume events quickly enough. Now we could (somewhat) protect against that by introducing a worker Go routine and another tracking data structure as suggested by @Stebalien in https://github.com/libp2p/go-libp2p/pull/1574#issuecomment-1143048078, but that data structure would need to be limited in size _somehow_, so this is no better than using a larger channel for the event bus subscriber. Using the new event bus dashboard (#2038) it should be easy to detect a misconfiguration there.

I introduced the change in the last commit. @MarcoPolo, @Stebalien does this reasoning make sense to you?

MarcoPolo commented on p2p/net/swarm/swarm_event_test.go:
```diff
@@ -0,0 +1,66 @@
+	select {
+	case <-sub.Out():
+		t.Fatal("didn't expect any more events")
+	case <-time.After(50 * time.Millisecond):
```
probably bump this to 100ms for CI. I know it'll slow things down 50ms, sorry.

marten-seemann replied:
Done.

MarcoPolo reviewed (approved):
Changes look good and reasoning makes sense.

> Unfortunately, I can't find a way to test the deduplication (that no event is emitted when a second connection is dialed)

You could try this:
Make a new test transport that wraps an existing transport. Put two multiaddrs for this test transport (they might have to be slightly different if we're deduping these somewhere). In the test transport have them synchronize with a waitgroup to return a dial at the same time.

marten-seemann commented:
> Make a new test transport that wraps an existing transport. Put two multiaddrs for this test transport (they might have to be slightly different if we're deduping these somewhere). In the test transport have them synchronize with a waitgroup to return a dial at the same time.

That doesn't work either. The dial worker loop _somewhere_ deduplicates. I have intention of diving into that mess to figure out where.
